### ISSUE TITLE
repair_additional_test.py::TestRepairAdditional::test_repair_kill_3 leaves a coredump

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#19244]
```
   @pytest.fixture(scope="function", autouse=False)
    def fixture_dtest_setup(  # noqa: PLR0912, PLR0913
 ...
    def check_errors_all_nodes(self, nodes=None, exclude_errors=None, search_str=None, regex=False):
        if nodes is None:
            nodes = self.cluster.nodelist()
...
        found_cores, ignored_cores = self.find_cores()
        if found_cores:
>           raise AssertionError("Core file(s) found. Marking test as failed.")
E           AssertionError: Core file(s) found. Marking test as failed.

dtest_setup.py:640: AssertionError
```
[dtest-gw4.log](https://github.com/user-attachments/files/15796022/dtest-gw4.log)
[1718138348177_repair_additional_test.py--TestRepairAdditional--test_repair_kill_3.zip](https://github.com/user-attachments/files/15796023/1718138348177_repair_additional_test.py--TestRepairAdditional--test_repair_kill_3.zip)

see https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2016/testReport/junit/repair_additional_test/TestRepairAdditional/FullDtest___full_split000___test_repair_kill_3/

**Comment by mykaul** [comment]
Is that on shutdown? All I see in the logs is:
```
WARN  2024-06-11 20:36:36,768 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: user-requested repair failed: std::runtime_error (["shard 0: std::runtime_error (repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)", "shard 1: std::runtime_error (repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)"])
WARN  2024-06-11 20:36:36,771 [shard 0: gms] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
INFO  2024-06-11 20:36:36,775 [shard 0: gms] raft - [aea4fbd5-d458-4f2d-beaa-e62b8f70a0b0] applier fiber stopped because state machine was aborted: abort requested
WARN  2024-06-11 20:36:37,371 [shard 0: gms] gossip - === Gossip round FAIL: seastar::abort_requested_exception (abort requested)
INFO  2024-06-11 20:36:37,371 [shard 0: gms] gossip - Gossip loop is not scheduled because it is disabled
WARN  2024-06-11 20:36:37,372 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:39,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:37:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:38:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
```

**Comment by Annamikhlin** [comment]
seen the same here https://jenkins.scylladb.com/job/scylla-enterprise/job/next/3016/testReport/junit/repair_additional_test/TestRepairAdditional/Build___x86___dtest_with_topology_changes___test_repair_kill_3/

logs:
[dtest-gw3.log](https://github.com/user-attachments/files/16123644/dtest-gw3.log)
[1720378470410_repair_additional_test.py__TestRepairAdditional__test_repair_kill_3.zip](https://github.com/user-attachments/files/16123646/1720378470410_repair_additional_test.py__TestRepairAdditional__test_repair_kill_3.zip)

**Comment by asias** [comment]
It looks like the problem is with shutdown. The repair has stopped successfully due to the shutdown. 

```
INFO  2024-06-11 20:36:36,759 [shard 0:main] init - Shutting down transport controller API was successful
INFO  2024-06-11 20:36:36,759 [shard 0:main] init - Shutting down local storage
INFO  2024-06-11 20:36:36,759 [shard 0:main] storage_service - Stop transport: starts
INFO  2024-06-11 20:36:36,759 [shard 0:main] migration_manager - stopping migration service
INFO  2024-06-11 20:36:36,760 [shard 1:main] migration_manager - stopping migration service
INFO  2024-06-11 20:36:36,766 [shard 0: gms] raft_topology - updating topology state: Starting new topology coordinator aea4fbd5-d458-4f2d-beaa-e62b8f70a0b0
INFO  2024-06-11 20:36:36,767 [shard 1:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: stats: repair_reason=repair, keyspace=ks, tables=["cf"], ranges_nr=513, round_nr=0, round_nr_fast_path_already_synced=0, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=4, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.009123228 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={}, row_from_disk_nr={}, row_from_disk_bytes_per_sec={} MiB/s, row_from_disk_rows_per_sec={} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
WARN  2024-06-11 20:36:36,767 [shard 1:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=["cf"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A 
INFO  2024-06-11 20:36:36,767 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: stats: repair_reason=repair, keyspace=ks, tables=["cf"], ranges_nr=513, round_nr=0, round_nr_fast_path_already_synced=0, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=4, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.009665471 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={}, row_from_disk_nr={}, row_from_disk_bytes_per_sec={} MiB/s, row_from_disk_rows_per_sec={} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
WARN  2024-06-11 20:36:36,767 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=["cf"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A 
INFO  2024-06-11 20:36:36,768 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: Started to shutdown off-strategy compaction updater
INFO  2024-06-11 20:36:36,768 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: Finished to shutdown off-strategy compaction updater
WARN  2024-06-11 20:36:36,768 [shard 0:strm] repair - repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: user-requested repair failed: std::runtime_error (["shard 0: std::runtime_error (repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)", "shard 1: std::runtime_error (repair[b481fe5f-e59c-4832-bd56-fc7a81ce5136]: 1 out of 513 ranges failed, keyspace=ks, tables=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)"])
WARN  2024-06-11 20:36:36,771 [shard 0: gms] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
INFO  2024-06-11 20:36:36,775 [shard 0: gms] raft - [aea4fbd5-d458-4f2d-beaa-e62b8f70a0b0] applier fiber stopped because state machine was aborted: abort requested
WARN  2024-06-11 20:36:37,371 [shard 0: gms] gossip - === Gossip round FAIL: seastar::abort_requested_exception (abort requested)
INFO  2024-06-11 20:36:37,371 [shard 0: gms] gossip - Gossip loop is not scheduled because it is disabled
WARN  2024-06-11 20:36:37,372 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:36:39,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:37:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
WARN  2024-06-11 20:38:37,373 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
```

The last message we saw is gossip failed to apply states. The local storage was shutdown. Perhaps gossip on_change listener accessed local storage while it was shutdown. 

BTW, the core and build is dev. Could not get a backtrace on the core.

**Comment by Annamikhlin** [comment]
seen again on https://jenkins.scylladb.com/job/scylla-master/job/next/7971/testReport/junit/repair_additional_test/TestRepairAdditional/Build___x86___dtest_with_topology_changes___test_repair_kill_3/


[dtest-gw4.log](https://github.com/user-attachments/files/16325012/dtest-gw4.log)
[1721568953705_repair_additional_test.py__TestRepairAdditional__test_repair_kill_3.zip](https://github.com/user-attachments/files/16325014/1721568953705_repair_additional_test.py__TestRepairAdditional__test_repair_kill_3.zip)

**Comment by denesb** [comment]
The coredump is generated by `SIGQUIT`, which is used by CCM to kill the node, when the wait for node shutdown times out.

**Comment by denesb** [comment]
On shard0 I can see a bunch of `defer()` tasks still in memory. There is also a `stop_transport()` in the `scylla task_histogram -a` output.

**Comment by denesb** [comment]
`stop_transport()`:
```
2824 future<> storage_service::stop_transport() {                                                                                                                                                                                                              
   1     if (!_transport_stopped.has_value()) {                                                                                                                                                                                                                
   2         promise<> stopped;                                                                                                                                                                                                                                
   3         _transport_stopped = stopped.get_future();                                                                                                                                                                                                        
   4                                                                                                                                                                                                                                                           
   5         seastar::async([this] {                                                                                                                                                                                                                           
   6             slogger.info("Stop transport: starts");                                                                                                                                                                                                       
   7                                                                                                                                                                                                                                                           
   8             slogger.debug("shutting down migration manager");                                                                                                                                                                                             
   9             _migration_manager.invoke_on_all(&service::migration_manager::drain).get();                                                                                                                                                                   
  10                                                                                                                                                                                                                                                           
  11             shutdown_protocol_servers().get();                                                                                                                                                                                                            
  12             slogger.info("Stop transport: shutdown rpc and cql server done");                                                                                                                                                                             
  13                                                                                                                                                                                                                                                           
  14             _gossiper.container().invoke_on_all(&gms::gossiper::shutdown).get();                                                                                                                                                                          
  15             slogger.info("Stop transport: stop_gossiping done");                                                                                                                                                                                          
  16                                                                                                                                                                                                                                                           
  17             _messaging.invoke_on_all(&netw::messaging_service::shutdown).get();                                                                                                                                                                           
  18             slogger.info("Stop transport: shutdown messaging_service done");                                                                                                                                                                              
  19                                                                                                                                                                                                                                                           
  20             _stream_manager.invoke_on_all(&streaming::stream_manager::shutdown).get();                                                                                                                                                                    
  21             slogger.info("Stop transport: shutdown stream_manager done");                                                                                                                                                                                 
  22                                                                                                                                                                                                                                                           
  23             slogger.info("Stop transport: done");                                                                                                                                                                                                         
  24         }).forward_to(std::move(stopped));                                                                                                                                                                                                                
  25     }                                                                                                                                                                                                                                                     
  26                                                                                                                                                                                                                                                           
  27     return _transport_stopped.value();                                                                                                                                                                                                                    
  28 } 
```
I see only the first log in the logs:
```
1153 INFO  2024-07-21 13:33:22,196 [shard 0:main] storage_service - Stop transport: starts                                                                                                                                                                     
```

**Comment by denesb** [comment]
Default stop timeout is 127 seconds, we can see this in action:
```
55  13:35:29,207 706     ccm                            WARNING  cluster.py          :760  | test_repair_kill_3: node1: node1 is still running after 127 seconds. Trying to generate coredump using kill(127209, SIGQUIT)...                                   
```

Node logs between initiating stop and the node being killed:
```
INFO  2024-07-21 13:33:22,195 [shard 0:main] init - Signal received; shutting down                                                                                                                                                                        
   1 INFO  2024-07-21 13:33:22,195 [shard 0:main] init - Shutting down thrift controller API                                                                                                                                                                   
   2 INFO  2024-07-21 13:33:22,195 [shard 0:main] compaction_manager - Stopped                                                                                                                                                                                 
   3 INFO  2024-07-21 13:33:22,195 [shard 0:strm] view_update_generator - leaving 0 unstaged sstables unprocessed                                                                                                                                              
   4 WARN  2024-07-21 13:33:22,195 [shard 0:strm] gossip - failure_detector_loop: Got error in the loop, live_nodes={127.0.76.2}: seastar::sleep_aborted (Sleep is aborted)                                                                                    
   5 INFO  2024-07-21 13:33:22,195 [shard 0:strm] gossip - failure_detector_loop: Finished main loop                                                                                                                                                           
   6 INFO  2024-07-21 13:33:22,195 [shard 1:main] compaction_manager - Asked to stop                                                                                                                                                                           
   7 INFO  2024-07-21 13:33:22,195 [shard 1:main] view_update_generator - Terminating background fiber                                                                                                                                                         
   8 INFO  2024-07-21 13:33:22,195 [shard 1:main] compaction_manager - Stopped                                                                                                                                                                                 
   9 INFO  2024-07-21 13:33:22,195 [shard 1:strm] view_update_generator - leaving 0 unstaged sstables unprocessed                                                                                                                                              
  10 INFO  2024-07-21 13:33:22,195 [shard 0:main] init - Shutting down thrift controller API was successful                                                                                                                                                    
  11 INFO  2024-07-21 13:33:22,195 [shard 0:main] init - Shutting down transport controller API                                                                                                                                                                
  12 WARN  2024-07-21 13:33:22,196 [shard 1:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: Repair job aborted by user, job=7347374d-0730-474d-aebf-6e5b4b9563ee, keyspace=ks, tables=["cf"]                                                      
  13 WARN  2024-07-21 13:33:22,196 [shard 1:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: shard=1, keyspace=ks, cf=cf, range=(-9197503542751403751,-9171109027002054351], got error in row level repair: seastar::abort_requested_exception (abo
     rt requested)                                                                                                                                                                                                                                             
  14 INFO  2024-07-21 13:33:22,196 [shard 0:main] init - Shutting down transport controller API was successful                                                                                                                                                 
  15 INFO  2024-07-21 13:33:22,196 [shard 0:main] init - Shutting down local storage                                                                                                                                                                           
  16 INFO  2024-07-21 13:33:22,196 [shard 0:main] storage_service - Stop transport: starts                                                                                                                                                                     
  17 INFO  2024-07-21 13:33:22,196 [shard 0:main] migration_manager - stopping migration service                                                                                                                                                               
  18 INFO  2024-07-21 13:33:22,196 [shard 1:main] migration_manager - stopping migration service                                                                                                                                                               
  19 WARN  2024-07-21 13:33:22,198 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: Repair job aborted by user, job=7347374d-0730-474d-aebf-6e5b4b9563ee, keyspace=ks, tables=["cf"]                                                      
  20 INFO  2024-07-21 13:33:22,200 [shard 1:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: stats: repair_reason=repair, keyspace=ks, tables=["cf"], ranges_nr=513, round_nr=2, round_nr_fast_path_already_synced=1, round_nr_fast_path_same_combi
     ned_hashes=0, round_nr_slow_path=0, rpc_call_nr=10, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.005214175 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={127.0.76.1: 324, 127.0.76.2: 324}, row_from_disk_nr={127.     0.76.1: 1, 127.0.76.2: 1}, row_from_disk_bytes_per_sec={127.0.76.1: 0.059259705, 127.0.76.2: 0.059259705} MiB/s, row_from_disk_rows_per_sec={127.0.76.1: 191.78488, 127.0.76.2: 191.78488} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}                   
  21 WARN  2024-07-21 13:33:22,200 [shard 1:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: 1 out of 513 ranges failed, keyspace=ks, tables=["cf"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A   
  22 INFO  2024-07-21 13:33:22,202 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: stats: repair_reason=repair, keyspace=ks, tables=["cf"], ranges_nr=513, round_nr=2, round_nr_fast_path_already_synced=1, round_nr_fast_path_same_combi
     ned_hashes=0, round_nr_slow_path=1, rpc_call_nr=8, tx_hashes_nr=0, rx_hashes_nr=1, duration=0.00776742 seconds, tx_row_nr=0, rx_row_nr=1, tx_row_bytes=0, rx_row_bytes=564, row_from_disk_bytes={127.0.76.1: 0, 127.0.76.2: 324}, row_from_disk_nr={127.0.     76.1: 0, 127.0.76.2: 1}, row_from_disk_bytes_per_sec={127.0.76.1: 0, 127.0.76.2: 0.039780322} MiB/s, row_from_disk_rows_per_sec={127.0.76.1: 0, 127.0.76.2: 128.74287} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={127.0.76.2: 1}                          
  23 WARN  2024-07-21 13:33:22,202 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: 0 out of 513 ranges failed, keyspace=ks, tables=["cf"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A   
  24 INFO  2024-07-21 13:33:22,202 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: Started to shutdown off-strategy compaction updater                                                                                                   
  25 INFO  2024-07-21 13:33:22,202 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: Finished to shutdown off-strategy compaction updater                                                                                                  
  26 WARN  2024-07-21 13:33:22,202 [shard 0:strm] repair - repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: user-requested repair failed: std::runtime_error (["shard 0: std::runtime_error (repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: 0 out of 513 ranges fai
     led, keyspace=ks, tables=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)", "shard 1: std::runtime_error (repair[7347374d-0730-474d-aebf-6e5b4b9563ee]: 1 out of 513 ranges failed, keyspace=ks, tab     les=[\"cf\"], repair_reason=repair, nodes_down_during_repair={}, aborted_by_user=true, failed_because=N/A)"])                                                                                                                                             
  27 INFO  2024-07-21 13:33:22,208 [shard 0: gms] raft_topology - updating topology state: Starting new topology coordinator 07c97cff-4258-4436-9bc9-37d4fd0795c0                                                                                              
  28 WARN  2024-07-21 13:33:22,211 [shard 0: gms] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
  29 INFO  2024-07-21 13:33:22,212 [shard 0: gms] raft - [07c97cff-4258-4436-9bc9-37d4fd0795c0] applier fiber stopped because state machine was aborted: abort requested                                                                                       
  30 WARN  2024-07-21 13:33:22,814 [shard 0: gms] gossip - === Gossip round FAIL: seastar::abort_requested_exception (abort requested)                                                                                                                         
  31 INFO  2024-07-21 13:33:22,814 [shard 0: gms] gossip - Gossip loop is not scheduled because it is disabled                                                                                                                                                 
  32 WARN  2024-07-21 13:33:23,616 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
  33 WARN  2024-07-21 13:33:23,616 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
  34 WARN  2024-07-21 13:33:25,617 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
  35 WARN  2024-07-21 13:34:23,616 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
  36 WARN  2024-07-21 13:35:23,617 [shard 0:main] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)
```

We don't see the logs from `shutdown_protocol_servers()`, so the part of `stop_transport()` that is stuck is `_migration_manager.invoke_on_all(&service::migration_manager::drain).get();`. We can see it was started, but unfortunately there is no log message when it is successful, so we can only infer it never finished.

**Comment by denesb** [comment]
Finally stumbled on the stuck fiber:
```
(gdb) scylla fiber 0x600004365080

[shard  1] #-7 (task*) 0x000060100466cb40 0x0000000002dcbff0 service::migration_manager::drain() [clone .resume] 
[shard  1] #-6 (task*) 0x000060100035cbc0 0x00000000006a9140 vtable for seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}, seastar::future<void>::then_wrapped_nrvo<void, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}>(seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void> + 16 
[shard  0] #-5 (task*) 0x000060000439cb60 0x00000000006a9028 vtable for seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}> + 16 
[shard  0] #-4 (task*) 0x00006000002f5f60 0x0000000000a4c700 vtable for seastar::parallel_for_each_state + 16 
[shard  0] #-3 (task*) 0x0000600004d9fea8 0x0000000000a4c4f8 vtable for seastar::(anonymous namespace)::thread_wake_task + 16 
[shard  0] #-2 (task*) 0x0000600004133380 0x0000000000a99078 vtable for seastar::thread_context + 16 
[shard  0] #-1 (task*) 0x00006000040b0f00 0x00000000007ad078 vtable for seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, service::storage_service::stop_transport()::$_0&&)::{lambda()#2}, seastar::future<void>::then_impl_nrvo<seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, service::storage_service::stop_transport()::$_0&&)::{lambda()#2}, seastar::future<void> >(service::storage_service::stop_transport()::$_0&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, auto:1&&, (auto:2&&)...)::{lambda()#2}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void> + 16 
[shard  0] #0  (task*) 0x0000600004365080 0x00000000007ad1f8 vtable for seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<void>::finally_body<seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, service::storage_service::stop_transport()::$_0&&)::{lambda()#3}, false>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, service::storage_service::stop_transport()::$_0&&)::{lambda()#3}, false> >(seastar::future<void>::finally_body<seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, service::storage_service::stop_transport()::$_0&&)::{lambda()#3}, false>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<seastar::async<service::storage_service::stop_transport()::$_0>(seastar::thread_attributes, auto:1&&, (auto:2&&)...)::{lambda()#3}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void> + 16 

Found no further pointers to task objects.
If you think there should be more, run `scylla fiber 0x0000600004365080 --verbose` to learn more.
Note that continuation across user-created seastar::promise<> objects are not detected by scylla-fiber.
```

**Comment by denesb** [comment]
And here is the rest of the fiber:
```
(gdb) scylla fiber 0x601004cbc730

[shard  1] #-1 (task*) 0x0000601004d3a100 0x0000000000554f18 vtable for seastar::continuation<seastar::internal::promise_base_with_type<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >, seastar::get_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>&, unsigned long)::{lambda()#1}, seastar::future<void>::then_impl_nrvo<seastar::get_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>&, unsigned long)::{lambda()#1}, seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> > >(seastar::get_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>&, unsigned long)::{lambda()#1}&&)::{lambda(seastar::internal::promise_base_with_type<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >&&, seastar::get_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<auto:1, auto:2>&, unsigned long)::{lambda()#1}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void> + 16 
[shard  1] #0  (task*) 0x0000601004cbc730 0x000000000059ced8 vtable for seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >::discard_result()::{lambda((auto:1&&)...)#1}, seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >::then_impl_nrvo<seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >::discard_result()::{lambda((auto:1&&)...)#1}, seastar::future<void> >(seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >::discard_result()::{lambda((auto:1&&)...)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >::discard_result()::{lambda((auto:1&&)...)#1}&, seastar::future_state<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >&&)#1}, seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> > + 16 
[shard  1] #1  (task*) 0x000060100466cb50 0x0000000000477540 vtable for seastar::internal::coroutine_traits_base<void>::promise_type + 16  (service::migration_manager::drain() at ./service/migration_manager.cc:81)
[shard  1] #2  (task*) 0x000060100035cbc0 0x00000000006a9140 vtable for seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}, seastar::future<void>::then_wrapped_nrvo<void, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}>(seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose()::{lambda(auto:1)#1}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void> + 16 
[shard  0] #3  (task*) 0x000060000439cb60 0x00000000006a9028 vtable for seastar::smp_message_queue::async_work_item<seastar::sharded<service::migration_manager>::invoke_on_all(seastar::smp_submit_to_options, std::function<seastar::future<void> (service::migration_manager&)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}> + 16 
[shard  0] #4  (task*) 0x0000600004c6da20 0x0000000002d7a5d0 service::migration_manager::drain() 

Found no further pointers to task objects.
If you think there should be more, run `scylla fiber 0x0000601004cbc730 --verbose` to learn more.
Note that continuation across user-created seastar::promise<> objects are not detected by scylla-fiber.
```

Looks like `scylla fiber` is unable to follow through coroutines. @michoecho did you have a fix for that, or am I remembering something else?

**Comment by denesb** [comment]
Both `migration_manager` instances, on both shards, have `0` count instances on the `_group0_barrier`'s semaphore:
```
(gdb) p $14->_group0_barrier._sem._count

$21 = 0
(gdb) p $5->_group0_barrier._sem._count

$22 = 0
```
This means that both of those are still busy, and `migration_manager::drain()` is waiting on `co_await _group0_barrier.join();`.
Could it be that group0 is already broken, by the `abort_source`?
```
   2 INFO  2024-07-21 13:33:22,208 [shard 0: gms] raft_topology - updating topology state: Starting new topology coordinator 07c97cff-4258-4436-9bc9-37d4fd0795c0                                                                                              
   1 WARN  2024-07-21 13:33:22,211 [shard 0: gms] gossip - Fail to apply application_state: seastar::abort_requested_exception (abort requested)                                                                                                               
1166 INFO  2024-07-21 13:33:22,212 [shard 0: gms] raft - [07c97cff-4258-4436-9bc9-37d4fd0795c0] applier fiber stopped because state machine was aborted: abort requested 
```
And waiters on this `_group0_barrier` are now stuck forever?

@kbr-scylla @gleb-cloudius

**Comment by michoecho** [comment]
> Looks like scylla fiber is unable to follow through coroutines. @michoecho did you have a fix for that, or am I remembering something else?

No, my `scylla fiber` change only made it print the names of the coroutines when walking forward. Before that change, you would only see `vtable for seastar::internal::coroutine_traits_base<void>::promise_type + 16` for that.

> Looks like scylla fiber is unable to follow through coroutines.

I think it *should* be able to follow through coroutines. Is that really what happened here, that it wasn't able to?

**Comment by gleb-cloudius** [comment]
> 
> This means that both of those are still busy, and `migration_manager::drain()` is waiting on `co_await _group0_barrier.join();`. Could it be that group0 is already broken, by the `abort_source`?
> 

`storage_service::stop_transport()` is called from `storage_service::.drain_on_shutdown()` which is called from `main()` before `group0_service.abort()`, so group0 suppose to be alive at this point. But may be without a quorum.

**Comment by denesb** [comment]
I do recall another instance where we've seen this in the past: shutdown stuck forever because group0 lost quorum (unfortunately I don't remember the exact issue).
@kbr-scylla do you have an issue tracking this problem?

**Comment by denesb** [comment]
> I think it _should_ be able to follow through coroutines. Is that really what happened here, that it wasn't able to?

It looks like so, see the trace at https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283947719, the last task is the `drain()` coroutine. We see in https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283988790 that the trace in fact continues beyond the coroutine.

**Comment by gleb-cloudius** [comment]
Raft group operations have a timeout now. I see that the one in drain suppose to run with default timeout of one minute. After the timeout we check it quorum is available and print a message if it is not.

**Comment by denesb** [comment]
> Raft group operations have a timeout now. I see that the one in drain suppose to run with default timeout of one minute. After the timeout we check it quorum is available and print a message if it is not.

CCM's timeout for shutdown is 127 seconds, but this involves more than just group0 drain. So we need to either reduce the group0 timeout, or increase the CCM one.
Alternatively, we should make the group0 timeout configurable and use a much shorter one in dtest (CCM).

**Comment by michoecho** [comment]
> > I think it _should_ be able to follow through coroutines. Is that really what happened here, that it wasn't able to?
> 
> It looks like so, see the trace at [#19244 (comment)](https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283947719), the last task is the `drain()` coroutine. We see in [#19244 (comment)](https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283988790) that the trace in fact continues beyond the coroutine.

@denesb But the second trace doesn't continue the first one. The `drain()` call in the first trace is  on shard 1, while the `drain()` call in the second trace is on shard 0. These are two different branches of the `invoke_on_all`.

**Comment by michoecho** [comment]
> > I think it _should_ be able to follow through coroutines. Is that really what happened here, that it wasn't able to?
> 
> It looks like so, see the trace at [#19244 (comment)](https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283947719), the last task is the `drain()` coroutine. We see in [#19244 (comment)](https://github.com/scylladb/scylladb/issues/19244#issuecomment-2283988790) that the trace in fact continues beyond the coroutine.

Right.

It seems that the problem is that `_walk_backward` assumes that tasks hold a pointer to the start (i.e. the first byte of the allocator slot) of their parent task, but in the case of coroutines that's not right because the pointer is to the `promise` inside the parent coroutine frame, not to the frame itself.

In other words, we need to add 16 bytes to the `ptr_meta.ptr` in this line https://github.com/scylladb/scylladb/blob/390e01673be1f365e42f826490ed95aeede1a259/scylla-gdb.py#L3979, if `ptr_meta.ptr` points to a coroutine frame.

**Comment by kbr-scylla** [comment]
@denesb this looks exactly like https://github.com/scylladb/scylladb/issues/19223, and was supposed to be fixed by https://github.com/scylladb/scylladb/pull/19860.

@denesb this should be assigned to my team. @emdroid please look into this as your next thing.

**Comment by kbr-scylla** [comment]
@denesb that last coredump you investigated -- what commit was it?

**Comment by kbr-scylla** [comment]
If it was 36b57f3432cb from this post: https://github.com/scylladb/scylladb/issues/19244#issuecomment-2241657247, then that's before the fix c05e077a1365f9c20a1ab060ff65b000f4fe3176 was merged.

I think we can close as duplicate of https://github.com/scylladb/scylladb/issues/19223, but @denesb please confirm.

**Comment by kbr-scylla** [comment]
(BTW relying on timeout would be masking the issue instead of fixing the root cause; https://github.com/scylladb/scylladb/commit/c05e077a1365f9c20a1ab060ff65b000f4fe3176 is the proper fix -- passing abort_sources to the right places)

**Comment by denesb** [comment]
> If it was [36b57f3](https://github.com/scylladb/scylladb/commit/36b57f3432cb34e20df27a7bc90ff16b4ecb043f) from this post: [#19244 (comment)](https://github.com/scylladb/scylladb/issues/19244#issuecomment-2241657247), then that's before the fix [c05e077](https://github.com/scylladb/scylladb/commit/c05e077a1365f9c20a1ab060ff65b000f4fe3176) was merged.
> 
> I think we can close as duplicate of #19223, but @denesb please confirm.

Thanks, let's close this.

Closing as duplicate of https://github.com/scylladb/scylladb/issues/19223.
